### PR TITLE
Drop scenarios

### PR DESCRIPTION
**Pull request by akrmn**
I think I managed to get rid of all the user-facing scenarios as well as a few internal ones. I think the most remarkable one is the example in the troubleshooting docs (current version: https://docs.daml.com/daml/troubleshooting.html#how-to-test-that-a-contract-is-visible-to-a-party), so please let me know if the wording of the new version could be better.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/lib/DA/Cli/Damlc.hs [hunk 1]
```diff
@@ -865,7 +865,10 @@ execDocTest opts files =
           -- This is horrible but we do not have a way to change the import paths in a running
           -- IdeState at the moment.
           pure $ nubOrd $ mapMaybe (uncurry moduleImportPath) (zip files' pmS)
-      opts <- pure opts { optImportPath = importPaths <> optImportPath opts, optHaddock = Haddock True }
+      opts <- pure opts
+        { optImportPath = importPaths <> optImportPath opts
+        , optHaddock = Haddock True
+        , optEnableScenarios = EnableScenarios True }
```
Comment (akrmn) [inline compiler/damlc/lib/DA/Cli/Damlc.hs:871]: This is a slightly different change: for doctests, we now enable scenarios implicitly since that's an implementation detail. Eventually we might want to change the implementation of `daml doctest` to use `Script` instead of `Scenario`, but that would mean requiring packages with doctests to depend on `daml-script` even if they don't define any scripts.

**Review by cocreature — APPROVED** [review]
Great work thanks a lot!

File: language-support/hs/bindings/examples/group-chat/daml/GroupChat.daml [hunk 2]
```diff
@@ -79,113 +81,125 @@ defaulting x y = Some (DA.Optional.fromOptional y x)
...
-    submitMustFail alice do fetch m3
-    submit         bob   do fetch m3
-    submit         chris do fetch m3
+template Helper
+  with
+    p : Party
+  where
+    signatory p
+
+    choice Fetch : Message
```
Comment (robin-da) [inline language-support/hs/bindings/examples/group-chat/daml/GroupChat.daml:90]: is this something you could have done with `queryContractId`?

File: docs/source/daml/code-snippets/Keys.daml [hunk 3]
```diff
@@ -94,177 +96,251 @@ template Delegation
       do
         fetchByKey @Keyed lookupKey
 
-lookupTest = scenario do
+template Helper
```
Comment (robin-da) [inline docs/source/daml/code-snippets/Keys.daml:99]: What's the idea of this Helper contract?

**Review by robin-da — APPROVED** [review]
Looks good!

File: daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs [hunk 4]
```diff
@@ -743,7 +743,7 @@ templateTests = testGroup "templates" $
             , ("gsg-trigger", [])
             -- daml-intro-1 - daml-intro-6 are not full projects.
             , ("daml-intro-7", [])
-            , ("daml-patterns", ["--enable-scenarios=yes"]) -- TODO: https://github.com/digital-asset/daml/issues/11316
```
[the remark this answers is not in the document]
Comment (akrmn) [reply daml-assistant/integration-tests/src/DA/Daml/Assistant/IntegrationTests.hs:746]: ah good point, thanks!

File: language-support/hs/bindings/examples/group-chat/daml/GroupChat.daml (hunk 2 above)
Comment (akrmn) [reply language-support/hs/bindings/examples/group-chat/daml/GroupChat.daml:90]: Good point! yes I could have done that in this case, I think I just went with the `Helper` out of habit after adapting another test which did need it. I can rewrite lines 112-122 to use `queryContractId` instead if you want

File: docs/source/daml/code-snippets/Keys.daml (hunk 3 above)
Comment (akrmn) [reply docs/source/daml/code-snippets/Keys.daml:99]: Context: 

> 
> The `submit` and `submitMustFail` functions are actually methods of the `HasSubmit` class.
> 
> ```haskell
> class HasSubmit m cmds where
>     submit         : HasCallStack => Party -> cmds a -> m a
>     submitMustFail : HasCallStack => Party -> cmds a -> m ()
> ```
> 
> And we have two relevant instances,
> 
> ```haskell
> instance HasSubmit Scenario Update
> instance HasSubmit Script   Commands
> ```
> 

Now, the important part is that `Commands` isn't quite as powerful (on its own) as `Update` - there are a bunch of functions with `Update a` return type that don't have an equivalent `Commands a` function. Crucially, `Update` has a `Monad`/`Action` instance, while `Commands` is only `Applicative`. 

For instance, there are a few tests where we do things like

```
-- Maintainer can fetch
  submit sig do
    (cid, keyed) <- fetchByKey @Keyed sig
    assert (keyedCid == cid)
```

When translating this to `Script` + `Commands`, we notice two things: 

1. There's no `Commands` equivalent to `fetchByKey` - though as you mentioned earlier we do have the somewhat similar `queryContractId : (Template t, IsParties p, HasCallStack) => p -> ContractId t -> Script (Optional t)`
2. The `do` block isn't going to work in a `Commands` context, since the second line (`assert ...`) depends  on the result of the first one (`(cid, keyed) <- ...`)

In this particular case, we can rewrite the test somewhat straightforwardly, by replacing `fetchByKey` with `queryContractId` and moving the `assertion` out of the `Update/Commands` block up to the `Scenario/Script` context:

```
-- Maintainer can fetch
  Some keyed <- queryContractId sig sig
  assert (keyedCid == cid)
```

However, this isn't always possible. In particular, if the original test had `submitMustFail` instead of `submit`, there's no way we can get the result of the `Update/Commands` block in the `Scenario/Script` context (since `submitMustFail : HasCallStack => Party -> cmds a -> m ()`)

The advice I got from @cocreature on this (and which I've since noticed in a few places of our codebase, as a sort of folk knowledge) was to move the `Update` body to a choice in a `Helper` template. This way, the `Update` itself needs no change at all. To _use_ the body, however, we need to summon a `Helper` contract and exercise the choice, but that's one of the things we _can_ do in `Commands` with `createAndExerciseCmd : (Template t, Choice t c r) => t -> c -> Commands r)`.


TL;DR: for some tests, we'd want to have a function `: Update a -> Commands a`. The closest thing we can have is physically moving the `Update a` code into a choice `C` of a helper template `T`, and then call `createAndExerciseCmd T C`.

File: language-support/hs/bindings/examples/group-chat/daml/GroupChat.daml (hunk 2 above)
Comment (robin-da) [reply language-support/hs/bindings/examples/group-chat/daml/GroupChat.daml:90]: not an important change, I was asking more for my understanding.

Comment (akrmn) [reply language-support/hs/bindings/examples/group-chat/daml/GroupChat.daml:90]: fair :)

File: docs/source/daml/code-snippets/Keys.daml (hunk 3 above)
Comment (robin-da) [reply docs/source/daml/code-snippets/Keys.daml:99]: cool, that makes sense.
